File: logic.zig
pub const Piece = struct {
    type: PieceType,
    player: Player,
};

pub const Player = enum { white, black };

pub const PieceType = enum {
    pawn,
    rook,
    knight,
    bishop,
    queen,
    king,
};

pub const Game = struct {
    pub const boardSize = 8;

    board: [boardSize][boardSize]?Piece,
    turn: Player,

    pub fn move(self: *Game, m: Move) MoveError!void {
        const piece = self.board[m.sy][m.sx] orelse {
            return MoveError.NoPieceAtSource;
        };

        if (piece.player != self.turn) {
            return MoveError.WrongPlayerTurn;
        }

        self.board[m.dy][m.dx] = piece;
        self.board[m.sy][m.sx] = null;

        self.turn = switch (self.turn) {
            .white => .black,
            .black => .white,
        };
    }
};

pub fn init() Game {
    const boardSize = Game.boardSize;

    var game = Game{
        .turn = Player.white, // White always begins
        .board = undefined,
    };

    var x: usize = 0;
    var y: usize = 0;
    // Initialize empty board
    while (y < boardSize) : (y += 1) {
        while (x < boardSize) : (x += 1) {
            game.board[y][x] = null;
        }
    }

    // Place pawns
    x = 0;
    while (x < boardSize) : (x += 1) {
        game.board[1][x] = Piece{ .player = .black, .type = .pawn };
        game.board[6][x] = Piece{ .player = .white, .type = .pawn };
    }

    // Place rooks
    game.board[0][0] = Piece{ .player = .black, .type = .rook };
    game.board[0][7] = Piece{ .player = .black, .type = .rook };
    game.board[7][0] = Piece{ .player = .white, .type = .rook };
    game.board[7][7] = Piece{ .player = .white, .type = .rook };

    // Place knights
    game.board[0][1] = Piece{ .player = .black, .type = .knight };
    game.board[0][6] = Piece{ .player = .black, .type = .knight };
    game.board[7][1] = Piece{ .player = .white, .type = .knight };
    game.board[7][6] = Piece{ .player = .white, .type = .knight };

    // Place bishops
    game.board[0][2] = Piece{ .player = .black, .type = .bishop };
    game.board[0][5] = Piece{ .player = .black, .type = .bishop };
    game.board[7][2] = Piece{ .player = .white, .type = .bishop };
    game.board[7][5] = Piece{ .player = .white, .type = .bishop };

    // Place queens
    game.board[0][3] = Piece{ .player = .black, .type = .queen };
    game.board[7][3] = Piece{ .player = .white, .type = .queen };

    // Place kings
    game.board[0][4] = Piece{ .player = .black, .type = .king };
    game.board[7][4] = Piece{ .player = .white, .type = .king };

    return game;
}

pub const MoveError = error{
    InvalidNotation,
    NoPieceAtSource,
    WrongPlayerTurn,
};

pub const Move = struct {
    sx: u8,
    sy: u8,
    dx: u8,
    dy: u8,

    pub fn parse(input: []const u8) MoveError!Move {
        if (input.len != 4) {
            return MoveError.InvalidNotation;
        }

        const sx = try parseFile(input[0]);
        const sy = try parseRank(input[1]);
        const dx = try parseFile(input[2]);
        const dy = try parseRank(input[3]);

        return Move{
            .sx = sx,
            .sy = sy,
            .dx = dx,
            .dy = dy,
        };
    }
};

/// Converts a file character (e.g., 'a' to 'h') to a column index (0-based).
fn parseFile(c: u8) MoveError!u8 {
    if (c >= 'a' and c <= 'h') {
        return c - 'a';
    }
    return MoveError.InvalidNotation;
}

/// Converts a rank character (e.g., '1' to '8') to a row index (0-based).
fn parseRank(c: u8) MoveError!u8 {
    if (c >= '1' and c <= '8') {
        return '8' - c;
    }
    return MoveError.InvalidNotation;
}

File: logic.c
    #include "logic.h"

#include <string.h>

void init_board(Piece board[BOARD_SIZE][BOARD_SIZE])
{
    // Initialize empty board
    for (int y = 0; y < BOARD_SIZE; y++)
    {
        for (int x = 0; x < BOARD_SIZE; x++)
        {
            board[y][x].type = EMPTY;
            board[y][x].player = NONE;
        }
    }

    // Place pawns
    for (int x = 0; x < BOARD_SIZE; x++)
    {
        board[1][x].type = PAWN;
        board[1][x].player = WHITE;
        board[6][x].type = PAWN;
        board[6][x].player = BLACK;
    }

    // Place rooks
    board[0][0].type = board[0][7].type = ROOK;
    board[0][0].player = board[0][7].player = WHITE;
    board[7][0].type = board[7][7].type = ROOK;
    board[7][0].player = board[7][7].player = BLACK;

    // Place knights
    board[0][1].type = board[0][6].type = KNIGHT;
    board[0][1].player = board[0][6].player = WHITE;
    board[7][1].type = board[7][6].type = KNIGHT;
    board[7][1].player = board[7][6].player = BLACK;

    // Place bishops
    board[0][2].type = board[0][5].type = BISHOP;
    board[0][2].player = board[0][5].player = WHITE;
    board[7][2].type = board[7][5].type = BISHOP;
    board[7][2].player = board[7][5].player = BLACK;

    // Place queens
    board[0][3].type = QUEEN;
    board[0][3].player = WHITE;
    board[7][3].type = QUEEN;
    board[7][3].player = BLACK;

    // Place kings
    board[0][4].type = KING;
    board[0][4].player = WHITE;
    board[7][4].type = KING;
    board[7][4].player = BLACK;
}

// Validate the move (simple version)
int is_valid_move(Piece board[BOARD_SIZE][BOARD_SIZE], int sx, int sy, int dx, int dy)
{
    // Add simple move validation for now (e.g., no out-of-bound or same-player destination)
    if (dx < 0 || dx >= BOARD_SIZE || dy < 0 || dy >= BOARD_SIZE)
        return 0;
    if (board[sy][sx].player == board[dy][dx].player)
        return 0; // Can't move to a spot with the same player

    // Add specific movement rules for each piece here (e.g., pawns move forward, knights move in L shape)

    return 1; // Accepting all moves for now
}

int move_piece(Piece board[BOARD_SIZE][BOARD_SIZE], int sx, int sy, int dx, int dy)
{
    if (is_valid_move(board, sx, sy, dx, dy))
    {
        board[dy][dx] = board[sy][sx];
        board[sy][sx].type = EMPTY;
        board[sy][sx].player = NONE;
        return 1;
    }
    return 0;
}

// Parse the user input (e.g., "e2e4") and convert it into board coordinates
void parse_input(char *input, int *sx, int *sy, int *dx, int *dy)
{
    if (strlen(input) != 4)
    {
        *sx = *sy = *dx = *dy = -1; // Invalid input
        return;
    }

    *sx = input[0] - 'a'; // 'a' -> 0, 'b' -> 1, ..., 'h' -> 7
    *sy = '8' - input[1]; // '8' -> 0, '7' -> 1, ..., '1' -> 7
    *dx = input[2] - 'a';
    *dy = '8' - input[3];
}